File: dotnet/src/Zora.Coins/ZoraApiException.cs
#nullable enable
using System;
using System.Net;

namespace Zora.Coins;

/// <summary>The Zora API answered with an HTTP error after retries were used up.</summary>
public sealed class ZoraApiException : Exception
{
    /// <summary>HTTP status.</summary>
    public HttpStatusCode StatusCode { get; }

    /// <summary>The endpoint, e.g. <c>/coin</c>.</summary>
    public string Path { get; }

    /// <summary>The API's machine-readable reason, where it gives one: /quote answers 422 with <c>LIQUIDITY</c>.</summary>
    public string? ErrorType { get; }

    /// <summary>The raw response body.</summary>
    public string Body { get; }

    /// <summary>Whether Zora refused the request for exceeding its rate limit. An API key raises it.</summary>
    public bool IsRateLimited => (int)StatusCode == 429;

    /// <summary>Whether a quote failed because the pool can't fill a trade that size. Try a smaller amount.</summary>
    public bool IsInsufficientLiquidity => ErrorType == "LIQUIDITY";

    internal ZoraApiException(HttpStatusCode status, string message, string path, string? errorType, string body)
        : base($"zora {path}: {(int)status} {message}")
    {
        StatusCode = status;
        Path = path;
        ErrorType = errorType;
        Body = body;
    }
}

File: java/src/main/java/io/github/pgalyen1987/zora/ZoraApiException.java
package io.github.pgalyen1987.zora;

/** The Zora API answered with an HTTP error after retries were used up. */
public final class ZoraApiException extends RuntimeException {
    private static final long serialVersionUID = 1L;
    private final int status;
    private final String path;
    private final String errorType;
    private final String body;

    ZoraApiException(int status, String message, String path, String errorType, String body) {
        super("zora " + path + ": " + status + " " + message);
        this.status = status;
        this.path = path;
        this.errorType = errorType;
        this.body = body;
    }

    /** HTTP status. */
    public int status() {
        return status;
    }

    /** The endpoint, e.g. {@code /coin}. */
    public String path() {
        return path;
    }

    /** The API's machine-readable reason, where it gives one: /quote answers 422 with {@code LIQUIDITY}. May be null. */
    public String errorType() {
        return errorType;
    }

    /** The raw response body. */
    public String body() {
        return body;
    }

    /** Whether Zora refused the request for exceeding its rate limit. An API key raises it. */
    public boolean isRateLimited() {
        return status == 429;
    }

    /** Whether a quote failed because the pool can't fill a trade that size. Try a smaller amount. */
    public boolean isInsufficientLiquidity() {
        return "LIQUIDITY".equals(errorType);
    }
}
